backoff.go

```diff
@@ -0,0 +1,71 @@
+package pubsub
+
+import (
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/peer"
+)
+
+const (
+	MinBackoffDelay   = 100 * time.Millisecond
+	MaxBackoffDelay   = 10 * time.Second
+	TimeToLive        = 10 * time.Minute
+	BackoffMultiplier = 2
+)
+
+type backoffHistory struct {
+	duration  time.Duration
+	lastTried time.Time
+}
+
+type backoff struct {
+	mu   sync.Mutex
+	info map[peer.ID]*backoffHistory
+	ct   int // size threshold that kicks off the cleaner
+}
+
+func newBackoff(sizeThreshold int) *backoff {
+	return &backoff{
+		mu:   sync.Mutex{},
+		ct:   sizeThreshold,
+		info: make(map[peer.ID]*backoffHistory),
+	}
+}
+
+func (b *backoff) updateAndGet(id peer.ID) time.Duration {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	h, ok := b.info[id]
+	if !ok || time.Since(h.lastTried) > TimeToLive {
+		// first request goes immediately.
+		h = &backoffHistory{
+			duration: time.Duration(0),
+		}
+	} else if h.duration < MinBackoffDelay {
+		h.duration = MinBackoffDelay
+	} else if h.duration < MaxBackoffDelay {
+		h.duration = time.Duration(BackoffMultiplier * h.duration)
+		if h.duration > MaxBackoffDelay || h.duration < 0 {
+			h.duration = MaxBackoffDelay
+		}
+	}
+
+	h.lastTried = time.Now()
+	b.info[id] = h
+
+	if len(b.info) > b.ct {
+		b.cleanup()
+	}
```

Comment:
nit: if we have a lot of peers (more than `b.ct`), then we will run `b.cleanup()` on every call to this function (which requires looping though the entire map), even when there is nothing to cleanup. Perhaps we need a better way to handle this?

For example, if we maintain an additional heap datastructure which is sorted by the expiration time, then at every call to the function we can just remove all the expired entries one by one and remove them from the map.

Will leave to @vyzo to decide whether this is necessary.

Reply:
Lets see if it is actually a problem in practice

Reply:
Actually it might be problematic, lets run a background goroutine that does it periodically (say once a minute).

Reply:
Added in https://github.com/libp2p/go-libp2p-pubsub/pull/483/commits/a9f4edfbfdf537be3fa16d4792e2a81b25dfefee